### PR TITLE
[mixbox] Add extractor for "mixbox.live"

### PR DESCRIPTION
**Pull request by pzhlkj6612**
For live on the homepage of "mixbox" (Japanese: ミックスボックス). https://mixbox.live/

**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

Hello! This is a simple extractor for [MixBox（ミックスボックス）](https://mixbox.live/). No account required.

#### Data

```js
...
LIVE_VIDEO_key: "04C8BEEEB092D9659F9EAC1574FEB12E63F2A1F3C797166028276AABF1360435",
LIVE_VIDEO_url: "https://xeqhc35g1h.execute-api.ap-northeast-1.amazonaws.com/default/0015MediaLiveApi/streamingurl",
...

```

#### Algorithm

```js
...
for (; ; ) switch (t.prev = t.next) {
  case 0:
    return (o = e.commit) ('clearUrl'),
    r = '',
    l = 'https://xeqhc35g1h.execute-api.ap-northeast-1.amazonaws.com/default/0015MediaLiveApi/streamingurl',
    '04C8BEEEB092D9659F9EAC1574FEB12E63F2A1F3C797166028276AABF1360435',
    10,
    h = n(625).randomBytes(10).toString('base64').substring(0, 10),
    (m = new d.a('SHA-256', 'TEXT')).update(''.concat('04C8BEEEB092D9659F9EAC1574FEB12E63F2A1F3C797166028276AABF1360435').concat(h)),
    t.next = 11,
    m.getHash('HEX');
  case 11:
    return r = t.sent,
    f = ''.concat(l, '?hash=').concat(h),
    t.next = 15,
    c.a.get(f, {
      headers: {
        Authorization: 'Bearer '.concat(r)
      }
    }).then((function (e) {
      o('setUrl', e.data.url)
    })).catch((function () {
      o('setUrl', null)
    }));
  case 15:
  case 'end':
    return t.stop()
}
...

```

### M3u8 playlist

```m3u8
#EXTM3U
#EXT-X-VERSION:3
#EXT-X-TARGETDURATION:12
#EXT-X-MEDIA-SEQUENCE:1191605
#EXT-X-DISCONTINUITY-SEQUENCE:2
#EXT-X-KEY:METHOD=AES-128,URI="https://d1v045aykfix0b.cloudfront.net/key/aes128.key",IV=0x00000000000000000000000000122EB5
#EXT-X-PROGRAM-DATE-TIME:2023-04-04T15:12:33.000Z
#EXTINF:12.00000,
Mixbox_1_1191605.ts
...

```

### Failed during download test

```log
> python test/test_download.py TestDownload.test_MixBoxHomepage
[debug] Loaded 1792 extractors 
[MixBox:homepage] Extracting URL: https://mixbox.live/ 
[MixBox:homepage] mixbox-live: Downloading webpage 
[MixBox:homepage] mixbox-live: Fetching asset js file 
[MixBox:homepage] mixbox-live: Fetching playlist json 
[MixBox:homepage] mixbox-live: Downloading m3u8 information 
[debug] Formats sorted by: hasvid, ie_pref, lang, quality, res, fps, hdr:12(7), vcodec:vp9.2(10), channels, acodec, filesize, fs_approx, tbr, vbr, abr, asr, proto, vext, aext, hasaud, source, id 
[info] mixbox-live: Downloading 1 format(s): 0 
[info] Writing video metadata as JSON to: test_MixBoxHomepage_mixbox-live.info.json 
[debug] Invoking ffmpeg downloader on "https://d1v045aykfix0b.cloudfront.net/live/Mixbox_1.m3u8?key=1bdd63d2d234562b945e7d53ea6aa58fc0516c4df5dfe02946671a5bdb5f3946" 
[download] Destination: test_MixBoxHomepage_mixbox-live.ts 
[debug] ffmpeg command line: ffmpeg -y -loglevel verbose -headers "User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/94.0.4606.81 Safari/537.36
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8
Accept-Language: en-us,en;q=0.5
Sec-Fetch-Mode: navigate
" -i "https://d1v045aykfix0b.cloudfront.net/live/Mixbox_1.m3u8?key=1bdd63d2d234562b945e7d53ea6aa58fc0516c4df5dfe02946671a5bdb5f3946" -c copy -fs 10241 -f mpegts "file:test_MixBoxHomepage_mixbox-live.ts.part" 
ffmpeg version n6.0-12-ga6dc92968a-20230404 Copyright (c) 2000-2023 the FFmpeg developers
  built with gcc 12.2.0 (crosstool-NG 1.25.0.152_89671bf)
  configuration: --prefix=/ffbuild/prefix --pkg-config-flags=--static --pkg-config=pkg-config --cross-prefix=x86_64-w64-mingw32- --arch=x86_64 --target-os=mingw32 --enable-gpl --enable-version3 --disable-debug --disable-w32threads --enable-pthreads 
--enable-iconv --enable-libxml2 --enable-zlib --enable-libfreetype --enable-libfribidi --enable-gmp --enable-lzma --enable-fontconfig --enable-libvorbis --enable-opencl --disable-libpulse --enable-libvmaf --disable-libxcb --disable-xlib --enable-amf --enable-libaom --enable-libaribb24 --enable-avisynth --enable-chromaprint --enable-libdav1d --enable-libdavs2 --disable-libfdk-aac --enable-ffnvcodec --enable-cuda-llvm --enable-frei0r --enable-libgme --enable-libkvazaar --enable-libass --enable-libbluray --enable-libjxl --enable-libmp3lame --enable-libopus --enable-librist --enable-libssh --enable-libtheora --enable-libvpx --enable-libwebp --enable-lv2 --disable-libmfx --enable-libvpl --enable-openal --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libopenh264 --enable-libopenjpeg --enable-libopenmpt --enable-librav1e --enable-librubberband --enable-schannel --enable-sdl2 --enable-libsoxr --enable-libsrt --enable-libsvtav1 --enable-libtwolame --enable-libuavs3d --disable-libdrm --disable-vaapi --enable-libvidstab --enable-vulkan --enable-libshaderc --enable-libplacebo --enable-libx264 --enable-libx265 --enable-libxavs2 --enable-libxvid --enable-libzimg --enable-libzvbi --extra-cflags=-DLIBTWOLAME_STATIC --extra-cxxflags= --extra-ldflags=-pthread --extra-ldexeflags= --extra-libs=-lgomp --extra-version=20230404
  libavutil      58.  2.100 / 58.  2.100
  libavcodec     60.  3.100 / 60.  3.100
  libavformat    60.  3.100 / 60.  3.100
  libavdevice    60.  1.100 / 60.  1.100
  libavfilter     9.  3.100 /  9.  3.100
  libswscale      7.  1.100 /  7.  1.100
  libswresample   4. 10.100 /  4. 10.100
  libpostproc    57.  1.100 / 57.  1.100
[tcp @ 000001d4e020d400] Starting connection attempt to 13.33.62.3 port 443
[tcp @ 000001d4e020d400] Successfully connected to 13.33.62.3 port 443
[tls @ 000001d4e020ad00] Server closed the connection
[hls @ 000001d4e020a200] Skip ('#EXT-X-VERSION:3')
[hls @ 000001d4e020a200] Skip ('#EXT-X-DISCONTINUITY-SEQUENCE:2')
[hls @ 000001d4e020a200] Skip ('#EXT-X-PROGRAM-DATE-TIME:2023-04-04T15:53:21.000Z')
[hls @ 000001d4e020a200] HLS request for url 'https://d1v045aykfix0b.cloudfront.net/live/Mixbox_1_1191816.ts', offset 0, playlist 0
[hls @ 000001d4e020a200] Opening 'https://d1v045aykfix0b.cloudfront.net/key/aes128.key' for reading
[tcp @ 000001d4e19d1680] Starting connection attempt to 13.33.62.3 port 443
[tcp @ 000001d4e19d1680] Successfully connected to 13.33.62.3 port 443
[AVIOContext @ 000001d4e19d1880] Statistics: 16 bytes read, 0 seeks
[hls @ 000001d4e020a200] Opening 'crypto+https://d1v045aykfix0b.cloudfront.net/live/Mixbox_1_1191816.ts' for reading
[tcp @ 000001d4e19d2140] Starting connection attempt to 13.33.62.3 port 443
[tcp @ 000001d4e19d2140] Successfully connected to 13.33.62.3 port 443
[h264 @ 000001d4e19fd680] Reinit context to 1280x720, pix_fmt: yuv420p
[mpegts @ 000001d4e020e980] parser not found for codec timed_id3, packets or times may be invalid.
Input #0, hls, from 'https://d1v045aykfix0b.cloudfront.net/live/Mixbox_1.m3u8?key=1bdd63d2d234562b945e7d53ea6aa58fc0516c4df5dfe02946671a5bdb5f3946':
  Duration: N/A, start: 80668.364356, bitrate: N/A
  Program 0
    Metadata:
      variant_bitrate : 0
  Stream #0:0: Video: h264 (Main), 1 reference frame ([27][0][0][0] / 0x001B), yuv420p(left), 1280x720 [SAR 1:1 DAR 16:9], 30 fps, 30 tbr, 90k tbn
    Metadata:
      variant_bitrate : 0
  Stream #0:1: Audio: aac (LC) ([15][0][0][0] / 0x000F), 48000 Hz, stereo, fltp
    Metadata:
      variant_bitrate : 0
  Stream #0:2: Data: timed_id3 (ID3  / 0x20334449)
    Metadata:
      variant_bitrate : 0
[mpegts @ 000001d4e1a40e40] service 1 using PCR in pid=256, pcr_period=100ms
[mpegts @ 000001d4e1a40e40] muxrate VBR, sdt every 500 ms, pat/pmt every 100 ms
Output #0, mpegts, to 'file:test_MixBoxHomepage_mixbox-live.ts.part':
  Metadata:
    encoder         : Lavf60.3.100
  Stream #0:0: Video: h264 (Main), 1 reference frame ([27][0][0][0] / 0x001B), yuv420p(left), 1280x720 (0x0) [SAR 1:1 DAR 16:9], q=2-31, 30 fps, 30 tbr, 90k tbn
    Metadata:
      variant_bitrate : 0
  Stream #0:1: Audio: aac (LC) ([15][0][0][0] / 0x000F), 48000 Hz, stereo, fltp
    Metadata:
      variant_bitrate : 0
Stream mapping:
  Stream #0:0 -> #0:0 (copy)
  Stream #0:1 -> #0:1 (copy)
Press [q] to stop, [?] for help
[out#0/mpegts @ 000001d4e1a33140] Error muxing a packet00.03 bitrate=  -0.0kbits/s speed=N/A
[out#0/mpegts @ 000001d4e1a33140] Terminating muxer thread
No more output streams to write to, finishing.
[AVIOContext @ 000001d4e247bd00] Statistics: 348740 bytes written, 0 seeks, 2 writeouts
frame=   24 fps=0.0 q=-1.0 Lsize=     341kB time=00:00:00.76 bitrate=3639.0kbits/s speed=18.6x    
video:327kB audio:0kB subtitle:0kB other streams:0kB global headers:0kB muxing overhead: 3.993153%
Input file #0 (https://d1v045aykfix0b.cloudfront.net/live/Mixbox_1.m3u8?key=1bdd63d2d234562b945e7d53ea6aa58fc0516c4df5dfe02946671a5bdb5f3946):
  Input stream #0:0 (video): 25 packets read (336310 bytes);
  Input stream #0:1 (audio): 12 packets read (6049 bytes); 
  Input stream #0:2 (data): 1 packets read (26 bytes);
  Total: 38 packets (342385 bytes) demuxed
Output file #0 (file:test_MixBoxHomepage_mixbox-live.ts.part):
  Output stream #0:0 (video): 24 packets muxed (334850 bytes);
  Output stream #0:1 (audio): 1 packets muxed (499 bytes);
  Total: 25 packets (335349 bytes) muxed
Terminating demuxer thread 0
[AVIOContext @ 000001d4e1a073c0] Statistics: 494160 bytes read, 0 seeks
[AVIOContext @ 000001d4e19c4dc0] Statistics: 1772 bytes read, 0 seeks
Conversion failed!
 

ERROR: ffmpeg exited with code 3753488571 

```

I'm using "ffmpeg-n6.0-12-ga6dc92968a-win64-gpl-6.0-yt-dlp.zip" of "Auto-Build 2023-04-04 12:54" from [yt-dlp/FFmpeg-Builds](https://github.com/yt-dlp/FFmpeg-Builds/releases).

Am I missing something?

<br>

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

**Comment by pzhlkj6612** [comment]
As of 1c250ad0bf4e677b979b390961e22d258e82d99b , this extractor supports downloading VoD on https://mixbox.live/.

### Analysis

#### Ticket code for `https://mixbox.live/campaign/\d{3,}` (useless)

> ## シリアルコードログイン  Serial code login
> 
> シリアルコードを入力して  Enter serial code
> ログインをクリック or タップしてください  Click or tap Login
> ※ ハイフン込み14ケタでご入力ください  Please enter in 14 digits including hyphens.
> 
> [ ABCD-1234-9876 ]
> 
> [ ログイン ]  Login

```js
  var n = r(101),
    o = r.n(n),
    i = window.siteEnv ? window.siteEnv : "v1",
    u = "https://7r8z5wp1de.execute-api.ap-northeast-1.amazonaws.com/".concat(i, "/authorization"),
    a = ("https://7r8z5wp1de.execute-api.ap-northeast-1.amazonaws.com/".concat(i, "/verification"), "https://0re84iq7jf.execute-api.ap-northeast-1.amazonaws.com/".concat(i, "/jwt/generate1"), ["1"]),
    c = /^MPZ/,
    s = "mixbox-".concat("2023040201");
  var f, l, h, p, v, d, y;
  f = "onmouseup" in document ? "click" : "touchend", l = document.querySelector("#ticket-login button"), h = document.getElementById("ticket"), p = document.getElementById("error"), v = window.navigator.userAgent, d = document.referrer, y = "", h.addEventListener("input", (function (t) {
    var e = h.value.toLocaleUpperCase();
    l.classList.add("button-disable"), c.test(e) && document.getElementById("button-send").classList.remove("button-disable")
  })), l.addEventListener(f, (function (t) {
    l.classList.contains("button-loading") || l.classList.contains("button-disable") || l.classList.contains("button-disable-force") || (y = h.value, p.innerHTML = "", 0 === y.length ? p.innerHTML = "<span>チケット番号を入力してください。</span>" : 14 === y.length ? function (t, e, r, n, i, a, c, f) {

```

It's server-side validation, so don't try to brute force the code. Just go to `https://mixbox.live/campaign_live/\d{3,}` .

#### Data in `https://mixbox.live/campaign_live/\d{3,}`

```js
data: function() {
    return {
        event: {
            title: "『ウマ娘 プリティーダービー WINNING LIVE 10』発売記念イベント",
            srcLive: "live046.html",
            srcArchive: "https://ext-npfbgn.b-ch.com/mixbox/54f01c289b5c60e1680230586d988de0872a5652",
            copyright: "© Cygames, Inc."
        },
        movieStatus: "live"
    }
},

// ...

checkArchiveTimer: function(t) {
    // ...
                    e.openArchiveTimer("2023/04/03 00:00", "2023/04/03 18:00", "2023/04/10 00:00", "2023/04/10 00:00", t);
                case 2:
                    e.movieStatus = n.sent;
                // ...

// ...

openArchiveTimer: function(e, t, n, r) {
    // ...
                    return n = e.t0,
                    o = c.a.tz(t, "Asia/Tokyo").utc(),
                    l = c.a.tz(r, "Asia/Tokyo").utc(),
                    h = "live",
                    n.isBetween(o, l) ? h = "archive" : n.isAfter(l) && (h = "end"),

```

### Query "playparam" (m3u8 playlist url and full title)

Data in `${srcArchive}`:

```html
<!-- ... -->
<video
<!-- ... -->
data-authtoken="f1cdc2ad98fe4295dd6c587760d7863b69e95115,1681040057,e34467e4ab5b778e1e746467e03afab28b04e8bd"
data-tenant_id="live2vod"
data-episode_id="54f01c289b5c60e1680230586d988de0872a5652"
<!-- ... -->
></video>
<!-- ... -->

```

playparam:

```json
{
    "header": //...
    "data": {
        "playparam": {
            "url": "https://npfbgn.b-ch.com/delivery/m3u8/playlist/cWV5SStHaXZKZjJXRXhBak1QWUpUUWdmZ3IwSkE0RktoenlrMjJSLzhmKzdlVUV5OGExU0VJRXBZSXhWY09oWVJQMDdWMU1EK0RSQy9EZmhhbjROdEVCcDFwTjdVcm9NcUlQd0ppaThLcStYUEg3QVEvc3RyZTAwTTNxNkFmbjRqQXRDNnZOYTk3Z00wY29QT2wwd2RDQWkrbDAwZlNxLzliWklXZ3p6enh0V292ZHhyWERLa29nb3RROEFwY0lj/index.m3u8"
        },
        "renditioninfo": [
            {
                "name": "240p",
                "video_width": 426,
                "video_height": 240,
                "video_bitrate": 236
            },
            // ...
        ],
        "thumbnailset": //...
        "subtitle": //...
        "playlog": {
            // ...
        },
        "display_title": "MixBox 『ウマ娘 プリティーダービー WINNING LIVE 10』発売記念イベント"
    }
}

```

EOF

File: yt_dlp/extractor/mixbox.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+            'title': 'MixBox',
+            'formats': formats,
+            'live_status': live_status,
+        }
+
+
+class MixBoxCampaignIE(InfoExtractor):
+    IE_NAME = 'MixBox:campaign'
+    IE_DESC = 'ミックスボックス：イベント'
+    _VALID_URL = r'https?://mixbox\.live/(?:campaign|campaign_live)/(?P<id>\d{3,})$'
```
Comment (Lesmiscore) [inline yt_dlp/extractor/mixbox.py:76]:
```suggestion
    _VALID_URL = r'https?://mixbox\.live/campaign(?:_live)?/(?P<id>\d+)$'
```

File: yt_dlp/extractor/mixbox.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+            'ext': 'ts',
+            'live_status': 'is_live',
+        },
+    }]
+
+    def _real_extract(self, url):
+        video_id = 'mixbox-live'
+
+        asset_js_urls = re.findall(
+            r'<script src="(/assets/.+?\.js)"', self._download_webpage(url, video_id))
```
Comment (Lesmiscore) [inline yt_dlp/extractor/mixbox.py:33]:
```suggestion
            r'<script src="(/assets/[^"]+?\.js)"', self._download_webpage(url, video_id))
```

**Review by Lesmiscore — APPROVED** [review]
totally looks okay to me

File: yt_dlp/extractor/mixbox.py [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    traverse_obj,
+    unified_timestamp,
+)
+from .common import InfoExtractor
+
+
+class MixBoxHomepageIE(InfoExtractor):
+    IE_NAME = 'MixBox:homepage'
+    IE_DESC = 'ミックスボックス'
+    _VALID_URL = r'https?://mixbox\.live/?$'
```
Comment (pukkandan) [inline yt_dlp/extractor/mixbox.py:18]:
```suggestion
    _VALID_URL = r'https?://mixbox\.live/?(?:[#?]|$)'
```

File: yt_dlp/extractor/mixbox.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+            asset_js_content = self._download_webpage(
+                f'https://mixbox.live/{asset_js_url}', video_id,
+                note='Fetching asset js file', errnote='Unable to fetch asset js file')
+            live_video_key, live_video_url = self._search_regex(
+                r'LIVE_VIDEO_key:\s*"(?P<key>[\dA-Z]{64})".+LIVE_VIDEO_url:\s*"(?P<url>https://[^"]+)"',
+                asset_js_content, name='live video info', group=('key', 'url'), default=(None, None))
+            if live_video_key and live_video_url:
+                break
+        if not live_video_key or not live_video_url:
+            raise ExtractorError('Cannot find live video info', expected=True)
```
Comment (pukkandan) [inline yt_dlp/extractor/mixbox.py:47]:
```suggestion

        for asset_js_url in reversed(asset_js_urls):
            asset_js_content = self._download_webpage(
                f'https://mixbox.live/{asset_js_url}', video_id,
                note='Fetching asset js file', errnote='Unable to fetch asset js file')
            live_video_key, live_video_url = self._search_regex(
                r'LIVE_VIDEO_key:\s*"(?P<key>[\dA-Z]{64})".+LIVE_VIDEO_url:\s*"(?P<url>https://[^"]+)"',
                asset_js_content, name='live video info', group=('key', 'url'), default=(None, None))
            if live_video_key and live_video_url:
                break
        else:
            raise ExtractorError('Cannot find live video info', expected=True)
```

**Review by pukkandan — CHANGES_REQUESTED** [review]

**Comment by pzhlkj6612** [comment]
According to the official notice <sup>[1]</sup>, mixbox.live has been already terminated on April 28th, 2023. This extractor is no longer useful.

Thank you all for your code review.

1: [音楽ストリーミングプラットフォーム「MixBox」サービス終了のお知らせ | News | Lantis web site](https://lantis.jp/news.php?id=2023041701) (Japanese)

**Pull request closed without merging**